wallet/src/types.rs

```diff
@@ -63,107 +63,102 @@ pub fn tx_fee(input_len: usize, output_len: usize, base_fee: Option<u64>) -> u64
+    NotEnoughFunds(u64),
 
-impl fmt::Display for Error {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		match *self {
-			_ => write!(f, "some kind of wallet error"),
-		}
-	}
-}
+    #[fail(display = "Fee dispute: sender fee {}, recipient fee {}", sender_fee, recipient_fee)]
+    FeeDispute{sender_fee: u64, recipient_fee: u64},
 
-impl From<keychain::Error> for Error {
-	fn from(e: keychain::Error) -> Error {
-		Error::Keychain(e)
-	}
-}
+    #[fail(display = "Fee exceeds amount: sender amount {}, recipient fee {}", sender_amount, recipient_fee)]
+    FeeExceedsAmount{sender_amount: u64,recipient_fee: u64},
 
-impl From<secp::Error> for Error {
-	fn from(e: secp::Error) -> Error {
-		Error::Secp(e)
-	}
-}
+    #[fail(display = "Keychain error")]
+    Keychain,
 
-impl From<transaction::Error> for Error {
-	fn from(e: transaction::Error) -> Error {
-		Error::Transaction(e)
-	}
-}
+    #[fail(display = "Transaction error")]
+    Transaction,
 
-impl From<serde_json::Error> for Error {
-	fn from(e: serde_json::Error) -> Error {
-		Error::Format(e.to_string())
-	}
-}
+    #[fail(display = "Secp error")]
+    Secp,
 
-// TODO - rethink this, would be nice not to have to worry about
-// low level hex conversion errors like this
-impl From<num::ParseIntError> for Error {
-	fn from(_: num::ParseIntError) -> Error {
-		Error::Format("Invalid hex".to_string())
-	}
+    #[fail(display = "Wallet data error: {}", _0)]
+    WalletData(&'static str),
+
+    /// An error in the format of the JSON structures exchanged by the wallet
+    #[fail(display = "JSON format error")]
+    Format,
+
+    #[fail(display = "I/O error")]
+    IO,
+
+    /// Error when contacting a node through its API
+    #[fail(display = "Node API error")]
+    Node,
+
+    /// Error originating from hyper.
+    #[fail(display = "Hyper error")]
+    Hyper,
+
+    /// Error originating from hyper uri parsing.
+    #[fail(display = "Uri parsing error")]
+    Uri,
+
+    #[fail(display = "Signature error")]
+    Signature(&'static str),
+
+    #[fail(display = "Generic error: {}", _0)]
+    GenericError(&'static str),
 }
 
-impl From<ser::Error> for Error {
-	fn from(e: ser::Error) -> Error {
-		Error::Format(e.to_string())
-	}
+
+impl Fail for Error {
+    fn cause(&self) -> Option<&Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
 }
 
-impl From<api::Error> for Error {
-	fn from(e: api::Error) -> Error {
-		Error::Node(e)
-	}
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Display::fmt(&self.inner, f)
+    }
 }
 
-impl From<io::Error> for Error {
-	fn from(e: io::Error) -> Error {
-		Error::IOError(e)
-	}
+impl Error {
+    pub fn kind(&self) -> ErrorKind {
+        *self.inner.get_context()
+    }
 }
 
-impl From<hyper::Error> for Error {
-	fn from(e: hyper::Error) -> Error {
-		Error::Hyper(e)
-	}
+impl From<ErrorKind> for Error {
```

Comment:
Oh I think I see the benefit here. Everything collapses into the single `Error` type and we just to be able to convert from ErrorKind to Error (once).